<a href="https://colab.research.google.com/github/gdkim0522/assign3_slime_AI/blob/main/Assignment3_Slime.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

In [241]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import sklearn.linear_model
from mpl_toolkits.mplot3d import Axes3D

import matplotlib.pyplot as plt
import numpy as np
from sklearn import datasets, linear_model
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.datasets import load_iris

#from IPython.display import display, Markdown, Latex
from IPython.display import Markdown as md

<span style='color:Black;font-size: 250%'># Part 1. Loading the dataset</span>

In [242]:
from sklearn.datasets import load_iris

#Load dataset remotely
iris = load_iris()
iris_df = pd.DataFrame(iris['data'], columns = iris.feature_names)
iris_df["species"] = iris.target


In [243]:
# Output the dataset
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [244]:
# Output the first 15 rows of the data
iris_df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    species  
0         0  
1         0  
2         0  
3         0  
4         0  
5         0  
6         0  
7         0  
8         0  
9         0  
10        0  
11        0  
12        0  
13        0  
14        0

In [245]:
# Display a summary of the table information (number of datapoints, etc.)
iris_df.info()
iris_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   species            150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB


sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)     species  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

In [246]:
iris_df.size

750

About the dataset
<br><br>
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

This is a dataset that explores relationship of the variables of the flower characteristics and type of species it is.

Our features are sepal length, sepal width, petal length, and petal width.   Our label is species.

Mapping is that label 0 is setosa, 1 is versicolor, 2 is virginica

<span style='color:Black;font-size: 250%'># Part 2. Splitting the dataset</span>

In [247]:
# Take the dataset and split it into our features (X) and label (y)
features = iris_df[["sepal length (cm)", "sepal width (cm)", "petal length (cm)", "petal width (cm)"]]
label = iris_df["species"]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, Y_train, Y_test = train_test_split(features, label,test_size=0.1, random_state = 25)


# print(X_train)
# print(X_test)
# print(Y_train)
# print(Y_test)
X_test

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
28                 5.2               3.4                1.4               0.2
72                 6.3               2.5                4.9               1.5
70                 5.9               3.2                4.8               1.8
85                 6.0               3.4                4.5               1.6
128                6.4               2.8                5.6               2.1
65                 6.7               3.1                4.4               1.4
127                6.1               3.0                4.9               1.8
42                 4.4               3.2                1.3               0.2
81                 5.5               2.4                3.7               1.0
93                 5.0               2.3                3.3               1.0
21                 5.1               3.7                1.5               0.4
31                 5.4               3.4                1.5               0.4
26                 5.0               3.4                1.6               0.4
133                6.3               2.8                5.1               1.5
6                  4.6               3.4                1.4               0.3

In [248]:
Y_test

28     0
72     1
70     1
85     1
128    2
65     1
127    2
42     0
81     1
93     1
21     0
31     0
26     0
133    2
6      0
Name: species, dtype: int32

<span style='color:Black;font-size: 250%'># Part 3. Perform Logistic Regression</span>

In [249]:
from sklearn.linear_model import LogisticRegression

# i. Use sklearn to train a LogisticRegression model on the training set
#log_regr = linear_model.LogisticRegression(max_iter = 500).fit(X_train, Y_train)

log_regr = LogisticRegression(random_state = 0, max_iter = 500).fit(X_train, Y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
X_test[:1]
log_regr.predict(X_test[:5])

array([0, 1, 2, 1, 2])

In [250]:
# ii. For a sample datapoint, predict the probabilities for each possible class
X_test[:1]
log_regr.predict(X_test[:5])

array([0, 1, 2, 1, 2])

In [251]:
log_regr.predict_proba(X_test[:5])

array([[9.74916920e-01, 2.50830477e-02, 3.19896491e-08],
       [8.77113902e-04, 6.01185725e-01, 3.97937161e-01],
       [2.61214863e-03, 4.09844071e-01, 5.87543780e-01],
       [1.11966605e-02, 7.58731568e-01, 2.30071772e-01],
       [1.42873879e-05, 3.76212502e-02, 9.62364462e-01]])

For each of the first five rows, it calculates the probabilities of what class the row belongs in.
The probability correspond to the prediction of the X_test dataset

In [252]:
# iii. Report on the score for Logistic regression model, what does the score measure?
log_regr.score(X_test, Y_test)

0.9333333333333333

The 0.9333 score value means there is an 93% chance of classification being correct.

In [253]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
# The coefficients
print("Coefficients: \n", log_regr.coef_)
# The intercept
print("Intercept: \n", log_regr.intercept_)

Coefficients: 
 [[-0.43642664  0.89470343 -2.43241387 -1.06251088]
 [ 0.56313925 -0.35330164 -0.18536135 -0.9834559 ]
 [-0.12671261 -0.54140179  2.61777522  2.04596679]]
Intercept: 
 [  9.80951197   2.03314201 -11.84265398]


<span style='color:Black;font-size: 250%'># Part 4. Support Vector Machine</span>

In [254]:
from sklearn import svm

# i. Use sklearn to train a Support Vector Classifier on the training set
svm_model = svm.SVC(probability=True)
#svm_model.fit(features, label)
svm_model.fit(X_train, Y_train)
X_train, Y_train
Y_test[:5]

28     0
72     1
70     1
85     1
128    2
Name: species, dtype: int32

In [255]:
# ii. For a sample datapoint, predict the probabilities for each possible class
#log_regr.predict_proba(X_test[:5])
svm_model.predict(X_test[:5])

array([0, 1, 1, 1, 2])

In [256]:
svm_model.predict_proba(X_test[:5])

array([[0.97275276, 0.01796723, 0.00928001],
       [0.00748191, 0.57406667, 0.41845142],
       [0.01062539, 0.44663728, 0.54273733],
       [0.0080946 , 0.90190308, 0.09000232],
       [0.00676047, 0.0044676 , 0.98877194]])

In [257]:
# iii. Report on the score for the SVM, what does the score measure?
svm_model.score(X_test, Y_test)

1.0

Using SVM, the score value is 1.0.
This means that SVM is perfect in predicitng the class of any data point feature.

<span style='color:Black;font-size: 250%'># Part 5. Neural Network</span>

In [258]:
from sklearn.neural_network import MLPClassifier
from sklearn.datasets import make_classification

#X, y = make_classification(n_samples=100, random_state=1)
#X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state=1)

# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
clf = MLPClassifier(random_state=1, max_iter=900).fit(X_train, Y_train)
#clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1, max_iter=900).fit(X_train, Y_train)

                                           
X_test[:1]

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
28                5.2               3.4                1.4               0.2

In [259]:
# ii. For a sample datapoint, predict the probabilities for each possible class
clf.predict_proba(X_test[:5])

array([[9.98687093e-01, 1.31290690e-03, 2.74724143e-14],
       [1.78199879e-04, 5.36492792e-01, 4.63329009e-01],
       [6.77371938e-04, 3.97944597e-01, 6.01378031e-01],
       [2.98350875e-03, 9.74627763e-01, 2.23887280e-02],
       [5.61076582e-07, 2.27386738e-03, 9.97725572e-01]])

In [260]:
clf.predict(X_test[:5])

array([0, 1, 2, 1, 2])

In [261]:
# iii. Report on the score for the Neural Network, what does the score measure?
clf.score(X_test, Y_test)

0.8666666666666667

This shows this type of classification is reliable, but not perfect.   It returns the mean accuracy on given data and labels

In [262]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)                                                     
clf = MLPClassifier(solver='lbfgs', alpha=0.5e-5, hidden_layer_sizes=(15,), random_state=1, max_iter=900).fit(X_train, Y_train)
clf.score(X_test, Y_test)

0.9333333333333333

This shows this type of model algorithm improves as different parameters are added and adjusted.  It improved the score from 0.86 to 0.93.   Different parameters that were adjusted include solver, alpha, and hidden layer size.

In [263]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)                                                     
clf = MLPClassifier(solver='sgd', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1, max_iter=500).fit(X_train, Y_train)
clf.score(X_test, Y_test)

C:\Users\gdkim\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


0.8666666666666667

<span style='color:Black;font-size: 250%'># Part 6: K-Nearest Neighbors</span>

In [264]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier(n_neighbors=50)
neigh.fit(X_train, Y_train)

KNeighborsClassifier(n_neighbors=50)

In [265]:
# ii. For a sample datapoint, predict the probabilities for each possible class
#print(neigh.predict([[1.1]]))
#print(neigh.predict_proba([[0.9]]))
neigh.predict(X_test[:5])

array([0, 1, 1, 1, 2])

In [266]:
neigh.predict_proba(X_test[:5])

array([[0.88, 0.12, 0.  ],
       [0.  , 0.58, 0.42],
       [0.  , 0.54, 0.46],
       [0.  , 0.68, 0.32],
       [0.  , 0.2 , 0.8 ]])

In [267]:
# iii. Report on the score for kNN, what does the score measure?
neigh.score(X_test, Y_test)

1.0

The score was 0.866 when neighbors=3, but it is one when neighbors=100   This shows this type of classification is reliable when small, but converges to perfection when neighbor size goes to infinity.   It returns the mean accuracy on given data and labels

<span style='color:Black;font-size: 250%'># Part 7: Conclusions and takeaways</span>
<br><br>
In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

The score value is 1, and it shows that polynomial regression of degree 2 is better than linear regression in terms of estimating values of scattered data points.
$h(x)= 2.06597033e^{-5} + 0.00000000e^{0}(1) +  1.20000000e^{1}(x_1) -1.32212252e^{-7}(x_2)+ 1.35861926e^{-11}(x_1^2) + 2.000000000e^{0}(x_1x_2) +  2.85714287e^{-2}(x_2^2) $
